crates/hir-def/src/visibility.rs

```diff
@@ -138,6 +138,12 @@ impl Visibility {
         // currently computing, so we must not call the `def_map` query for it.
         let def_map_block = def_map.block_id();
         loop {
+            // `to_module` is not a block module and should not be moved upward,
+            // as doing so might expose private items.
+            if !to_module.is_block_module() {
+                break;
+            }
+
             match (to_module.block, def_map_block) {
                 // to_module is not a block, so there is no parent def map to use
                 (None, _) => (),
```

Comment:
Wasn't the case for the your above code `if !to_module.is_block_module() { .. }` being handled here?

Reply:
Was wondering the same but actually not quite, `is_block_module` is defined as `self.block.is_some() && self.local_id == DefMap::ROOT`, so we are checking `self.block.is_none()` (the match arm part) or `self.local_id != ROOT` (thats the new thing).

So if I get this right, the problem is that if we are in a module within a block module we incorrectly move upwards out of the local module right now. (this is all so confusing ..,).

Can we encode that branch in the match as well (or clean up the branching in some other way)? This as proposed makes it difficult to read imo

Reply:
Thanks for pointing out.

Actually, the fix by checking whether `to_module` is a `block_module` was incorrect.

The problem was that we keep moving upwards even though we should stop as soon as we reach the same block as `from_module`'s `def_map`.